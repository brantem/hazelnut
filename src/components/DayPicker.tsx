import { ChangeEvent, useRef } from 'react';
import clsx from 'clsx';
import dayjs from 'dayjs';

import days from 'data/days';
import { Day } from 'types/shared';
import { getFirstDateDifferenceFromToday } from 'lib/helpers';

type DayProps = {
  day: string;
  isSelected: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  isDisabled?: boolean;
};

const Day = ({ day, isSelected, isDisabled, onChange }: DayProps) => {
  const inputRef = useRef<HTMLInputElement | null>(null);

  return (
    <label
      key={day}
      htmlFor={day}
      className={clsx(
        'flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2',
        {
          'border-black bg-black': isSelected,
          'bg-neutral-100': isDisabled,
          'cursor-pointer': !isDisabled,
          'text-white': isSelected,
          'text-neutral-500': isDisabled && !isSelected,
        },
      )}
      tabIndex={isDisabled ? -1 : 0}
      onKeyDown={(e) => !isDisabled && e.code === 'Space' && inputRef.current?.click()}
      data-testid={`day-picker-option-${day.toLowerCase()}`}
      aria-checked={isSelected ? 'true' : 'false'}
    >
      {day[0]}
      <input
        ref={inputRef}
        type="checkbox"
        id={day}
        name={day}
        checked={isSelected}
        onChange={onChange}
        className="sr-only"
        tabIndex={-1}
        disabled={isDisabled}
      />
    </label>
  );
};

type DayPickerProps = {
  value: Day[];
  onChange: (days: Day[]) => void;
  isDisabled?: boolean;
  showNext?: boolean;
};

const DayPicker = ({ value, onChange, isDisabled, showNext }: DayPickerProps) => {
  return (
    <div data-testid="day-picker">
      <label className="block text-sm font-medium text-neutral-700">Day(s)</label>

      <div className="mt-2 flex items-center justify-between">
        {days.map((day) => (
          <Day
            key={day}
            day={day}
            isSelected={value.includes(day)}
            onChange={(e) => onChange(e.target.checked ? [...value, day] : value.filter((_day) => _day !== day))}
            isDisabled={isDisabled}
          />
        ))}
      </div>

      {showNext && (
        <p className="mt-2 text-sm text-neutral-500">
          Next:{' '}
          {value.length
            ? dayjs().startOf('day').add(getFirstDateDifferenceFromToday(value), 'day').format('D MMM')
            : '-'}
        </p>
      )}
    </div>
  );
};

export default DayPicker;
