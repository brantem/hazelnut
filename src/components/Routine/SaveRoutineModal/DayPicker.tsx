import { ChangeEvent, useRef } from 'react';
import clsx from 'clsx';

import days from 'data/days';
import { Day as _Day } from 'types/shared';
import { sortDays } from 'lib/helpers';

type DayProps = {
  day: string;
  isSelected: boolean;
  onChange: (e: ChangeEvent<HTMLInputElement>) => void;
  isDisabled?: boolean;
};

const Day = ({ day, isSelected, isDisabled, onChange }: DayProps) => {
  const inputRef = useRef<HTMLInputElement | null>(null);

  return (
    <label
      key={day}
      htmlFor={day}
      className={clsx(
        'flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2',
        {
          'border-black bg-black': isSelected,
          'bg-neutral-100': isDisabled,
          'cursor-pointer': !isDisabled,
          'text-white': isSelected,
          'text-neutral-500': isDisabled && !isSelected,
        },
      )}
      tabIndex={isDisabled ? -1 : 0}
      onKeyDown={(e) => !isDisabled && e.code === 'Space' && inputRef.current?.click()}
      data-testid={`day-picker-option-${day.toLowerCase()}`}
      aria-checked={isSelected ? 'true' : 'false'}
    >
      {day[0]}
      <input
        ref={inputRef}
        type="checkbox"
        id={day}
        name={day}
        checked={isSelected}
        onChange={onChange}
        className="sr-only"
        tabIndex={-1}
        disabled={isDisabled}
      />
    </label>
  );
};

type DayPickerProps = {
  label?: string;
  value: _Day[];
  onChange: (days: _Day[]) => void;
  isDisabled?: boolean;
};

const DayPicker = ({ label = 'Day(s)', value, onChange, isDisabled }: DayPickerProps) => {
  return (
    <div data-testid="day-picker">
      {label && <label className="mb-1 block text-sm text-neutral-700">{label}</label>}

      <div className="mt-2 flex items-center justify-between">
        {days.map((day) => (
          <Day
            key={day}
            day={day}
            isSelected={value.includes(day)}
            onChange={(e) =>
              onChange(e.target.checked ? sortDays([...value, day]) : value.filter((_day) => _day !== day))
            }
            isDisabled={isDisabled}
          />
        ))}
      </div>
    </div>
  );
};

export default DayPicker;
