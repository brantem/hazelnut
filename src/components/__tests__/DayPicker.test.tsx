import { render, screen, act } from '@testing-library/react';

import '@testing-library/jest-dom';

import DayPicker from 'components/DayPicker';

describe('DayPicker', () => {
  it('should render successfully', () => {
    const onChange = vi.fn(() => {});
    const { container, rerender } = render(<DayPicker value={[]} onChange={onChange} />);

    expect(container).toMatchSnapshot();
    expect(container.querySelector('[aria-checked="true"]')).not.toBeInTheDocument();
    const option = screen.getByTestId('day-picker-option-monday');
    act(() => option.click());
    expect(onChange).toHaveBeenCalledWith(['MONDAY']);
    rerender(<DayPicker value={['MONDAY']} onChange={() => {}} />);
    expect(option).toHaveAttribute('aria-checked', 'true');
    expect(container).toMatchSnapshot();

    // test space
  });

  it("shouldn't be selectable when disabled", () => {
    const onChange = vi.fn(() => {});
    const { container } = render(<DayPicker value={[]} onChange={onChange} isDisabled />);

    expect(container).toMatchSnapshot();
    act(() => screen.getByTestId('day-picker-option-monday').click());
    expect(onChange).not.toHaveBeenCalled();

    // test space
  });
});
