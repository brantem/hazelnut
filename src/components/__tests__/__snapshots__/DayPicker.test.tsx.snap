// Vitest Snapshot v1

exports[`DayPicker > should render successfully 1`] = `
<div>
  <div
    data-testid="day-picker"
  >
    <label
      class="block text-sm font-medium text-neutral-700"
    >
      Day(s)
    </label>
    <div
      class="mt-2 flex items-center justify-between"
    >
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-monday"
        for="MONDAY"
        tabindex="0"
      >
        M
        <input
          class="sr-only"
          id="MONDAY"
          name="MONDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-tuesday"
        for="TUESDAY"
        tabindex="0"
      >
        T
        <input
          class="sr-only"
          id="TUESDAY"
          name="TUESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-wednesday"
        for="WEDNESDAY"
        tabindex="0"
      >
        W
        <input
          class="sr-only"
          id="WEDNESDAY"
          name="WEDNESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-thursday"
        for="THURSDAY"
        tabindex="0"
      >
        T
        <input
          class="sr-only"
          id="THURSDAY"
          name="THURSDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-friday"
        for="FRIDAY"
        tabindex="0"
      >
        F
        <input
          class="sr-only"
          id="FRIDAY"
          name="FRIDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-saturday"
        for="SATURDAY"
        tabindex="0"
      >
        S
        <input
          class="sr-only"
          id="SATURDAY"
          name="SATURDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-sunday"
        for="SUNDAY"
        tabindex="0"
      >
        S
        <input
          class="sr-only"
          id="SUNDAY"
          name="SUNDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`DayPicker > should render successfully 2`] = `
<div>
  <div
    data-testid="day-picker"
  >
    <label
      class="block text-sm font-medium text-neutral-700"
    >
      Day(s)
    </label>
    <div
      class="mt-2 flex items-center justify-between"
    >
      <label
        aria-checked="true"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 border-black bg-black cursor-pointer text-white"
        data-testid="day-picker-option-monday"
        for="MONDAY"
        tabindex="0"
      >
        M
        <input
          class="sr-only"
          id="MONDAY"
          name="MONDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-tuesday"
        for="TUESDAY"
        tabindex="0"
      >
        T
        <input
          class="sr-only"
          id="TUESDAY"
          name="TUESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-wednesday"
        for="WEDNESDAY"
        tabindex="0"
      >
        W
        <input
          class="sr-only"
          id="WEDNESDAY"
          name="WEDNESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-thursday"
        for="THURSDAY"
        tabindex="0"
      >
        T
        <input
          class="sr-only"
          id="THURSDAY"
          name="THURSDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-friday"
        for="FRIDAY"
        tabindex="0"
      >
        F
        <input
          class="sr-only"
          id="FRIDAY"
          name="FRIDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-saturday"
        for="SATURDAY"
        tabindex="0"
      >
        S
        <input
          class="sr-only"
          id="SATURDAY"
          name="SATURDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 cursor-pointer"
        data-testid="day-picker-option-sunday"
        for="SUNDAY"
        tabindex="0"
      >
        S
        <input
          class="sr-only"
          id="SUNDAY"
          name="SUNDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
    </div>
  </div>
</div>
`;

exports[`DayPicker > shouldn't be selectable when disabled 1`] = `
<div>
  <div
    data-testid="day-picker"
  >
    <label
      class="block text-sm font-medium text-neutral-700"
    >
      Day(s)
    </label>
    <div
      class="mt-2 flex items-center justify-between"
    >
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-monday"
        for="MONDAY"
        tabindex="-1"
      >
        M
        <input
          class="sr-only"
          disabled=""
          id="MONDAY"
          name="MONDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-tuesday"
        for="TUESDAY"
        tabindex="-1"
      >
        T
        <input
          class="sr-only"
          disabled=""
          id="TUESDAY"
          name="TUESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-wednesday"
        for="WEDNESDAY"
        tabindex="-1"
      >
        W
        <input
          class="sr-only"
          disabled=""
          id="WEDNESDAY"
          name="WEDNESDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-thursday"
        for="THURSDAY"
        tabindex="-1"
      >
        T
        <input
          class="sr-only"
          disabled=""
          id="THURSDAY"
          name="THURSDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-friday"
        for="FRIDAY"
        tabindex="-1"
      >
        F
        <input
          class="sr-only"
          disabled=""
          id="FRIDAY"
          name="FRIDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-saturday"
        for="SATURDAY"
        tabindex="-1"
      >
        S
        <input
          class="sr-only"
          disabled=""
          id="SATURDAY"
          name="SATURDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
      <label
        aria-checked="false"
        class="flex h-10 w-10 items-center justify-center rounded-full border text-center ring-neutral-500 ring-offset-2 focus-visible:outline-none focus-visible:ring-2 bg-neutral-100 text-neutral-500"
        data-testid="day-picker-option-sunday"
        for="SUNDAY"
        tabindex="-1"
      >
        S
        <input
          class="sr-only"
          disabled=""
          id="SUNDAY"
          name="SUNDAY"
          tabindex="-1"
          type="checkbox"
        />
      </label>
    </div>
  </div>
</div>
`;
